repair/row_level.cc

```diff
@@ -3826,3 +3826,75 @@ future<uint32_t> repair_service::get_next_repair_meta_id() {
 locator::host_id repair_service::my_host_id() const noexcept {
     return _gossiper.local().my_host_id();
 }
+
+// A struct to hold the first and last token of a tablet.
+struct tablet_token_range {
+    int64_t first_token;
+    int64_t last_token;
+};
+
+// Function to count the number of ranges in ranges1 covered by the merged ranges of ranges2.
+static size_t count_finished_tablets(utils::chunked_vector<tablet_token_range> ranges1, utils::chunked_vector<tablet_token_range> ranges2) {
+    if (ranges2.empty()) {
+        return 0;
+    }
+
+    // First, merge overlapping and adjacent ranges in ranges2.
+    auto& sorted = ranges2;
+    std::sort(sorted.begin(), sorted.end(), [](const auto& a, const auto& b) {
+        if (a.first_token != b.first_token) {
+            return a.first_token < b.first_token;
+        }
+        return a.last_token < b.last_token;
+    });
+
+    utils::chunked_vector<tablet_token_range> merged;
+    merged.push_back(sorted[0]);
+
+    for (size_t i = 1; i < sorted.size(); ++i) {
+        // To avoid overflow with max() + 1, we check adjacency with `a - 1 <= b` instead of `a <= b + 1`
+        if (sorted[i].first_token - 1 <= merged.back().last_token) {
+            merged.back().last_token = std::max(merged.back().last_token, sorted[i].last_token);
+        } else {
+            merged.push_back(sorted[i]);
+        }
+    }
+
+    // Now, count how many ranges in ranges1 are covered by the merged ranges.
+    size_t covered_count = 0;
+    for (const auto& range1 : ranges1) {
+        bool is_covered = std::any_of(merged.begin(), merged.end(), [&](const auto& r) {
+            return r.first_token <= range1.first_token && range1.last_token <= r.last_token;
+        });
```

Comment:
This has quadratic complexity. with 100'000 tablets that's too much.

Reply:
This is called on a slow path only when the progress is queried by the user. The number of merged ranges will reduce when more and  tablets finishes, e.g., the 100'000 will become 1 range when all finishes. 

Note, this is per table, not per whole cluster. If a single table has 100K tablets, this will translates to RF* 100K * *5GB/1024.0 = 1464 TB of data for a single table.